File: android/src/main/java/com/mrousavy/old/camera/frameprocessor/FrameProcessorPerformanceDataCollector.kt
package com.mrousavy.old.camera.frameprocessor

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.max
import kotlin.math.min

data class PerformanceSampleCollection(val endPerformanceSampleCollection: () -> Unit)

// keep a maximum of `maxSampleSize` historical performance data samples cached.
private const val maxSampleSize = 20
private const val warmupSamples = 5

class FrameProcessorPerformanceDataCollector {
  private var counter = AtomicInteger(0)
  private var performanceSamples: ArrayList<Double> = ArrayList(maxSampleSize)
  private val totalFramesProcessed = AtomicLong(0)
  private val totalProcessingTime = AtomicLong(0)
  private var minExecutionTime = Double.MAX_VALUE
  private var maxExecutionTime = Double.MIN_VALUE

  val averageExecutionTimeSeconds: Double
    get() = synchronized(performanceSamples) {
      if (performanceSamples.isEmpty()) return 0.0

      // Skip warmup samples for more accurate measurement
      val validSamples = if (performanceSamples.size > warmupSamples) {
        performanceSamples.drop(warmupSamples)
      } else {
        performanceSamples
      }

      if (validSamples.isEmpty()) return 0.0

      // Use trimmed mean to reduce impact of outliers
      val sorted = validSamples.sorted()
      val trimPercent = 0.1 // Remove top and bottom 10%
      val trimCount = (sorted.size * trimPercent).toInt()

      val trimmed = if (trimCount > 0 && sorted.size > trimCount * 2) {
        sorted.drop(trimCount).dropLast(trimCount)
      } else {
        sorted
      }

      return trimmed.average()
    }

  val hasEnoughData: Boolean
    get() = performanceSamples.size >= warmupSamples

  val processingStats: Map<String, Any>
    get() = mapOf(
      "totalFrames" to totalFramesProcessed.get(),
      "totalTime" to totalProcessingTime.get() / 1_000_000.0, // Convert to seconds
      "averageTime" to averageExecutionTimeSeconds,
      "minTime" to minExecutionTime,
      "maxTime" to maxExecutionTime,
      "sampleCount" to performanceSamples.size
    )

  fun beginPerformanceSampleCollection(): PerformanceSampleCollection {
    val begin = System.nanoTime()

    return PerformanceSampleCollection {
      val end = System.nanoTime()
      val nanoseconds = end - begin
      val seconds = nanoseconds / 1_000_000_000.0

      synchronized(performanceSamples) {
        val index = counter.getAndIncrement() % maxSampleSize

        if (performanceSamples.size > index) {
          performanceSamples[index] = seconds
        } else {
          performanceSamples.add(seconds)
        }

        // Update statistics
        totalFramesProcessed.incrementAndGet()
        totalProcessingTime.addAndGet(nanoseconds)
        minExecutionTime = min(minExecutionTime, seconds)
        maxExecutionTime = max(maxExecutionTime, seconds)
      }
    }
  }

  fun clear() {
    synchronized(performanceSamples) {
      counter.set(0)
      performanceSamples.clear()
      totalFramesProcessed.set(0)
      totalProcessingTime.set(0)
      minExecutionTime = Double.MAX_VALUE
      maxExecutionTime = Double.MIN_VALUE
    }
  }

  fun getSuggestedFrameRate(maxFrameRate: Double = 30.0): Double {
    if (!hasEnoughData) return maxFrameRate

    val avgTime = averageExecutionTimeSeconds
    if (avgTime <= 0) return maxFrameRate

    // Calculate theoretical max FPS based on processing time
    val theoreticalMaxFps = 1.0 / avgTime

    // Add safety margin (20%) to prevent frame drops
    val safeMaxFps = theoreticalMaxFps * 0.8

    // Clamp to reasonable bounds
    return min(safeMaxFps, maxFrameRate).coerceAtLeast(1.0)
  }
}

File: ios/Frame Processor/FrameProcessorPerformanceDataCollector.swift
//
//  FrameProcessorPerformanceDataCollector.swift
//  VisionCameraOld
//
//  Created by Marc Rousavy on 30.08.21.
//  Copyright © 2021 mrousavy. All rights reserved.
//

import Foundation
import os.log

// keep a maximum of `maxSampleSize` historical performance data samples cached.
private let maxSampleSize = 20
private let warmupSamples = 5

// MARK: - PerformanceSampleCollection

struct PerformanceSampleCollection {
  var endPerformanceSampleCollection: () -> Void

  init(end: @escaping () -> Void) {
    endPerformanceSampleCollection = end
  }
}

// MARK: - FrameProcessorPerformanceDataCollector

class FrameProcessorPerformanceDataCollector {
  private var performanceSamples: [Double] = []
  private let performanceSamplesQueue = DispatchQueue(label: "performance-samples", qos: .utility)
  private var counter = 0
  private var lastEvaluation = -1
  private var totalFramesProcessed: Int64 = 0
  private var totalProcessingTime: UInt64 = 0
  private var minExecutionTime: Double = Double.greatestFiniteMagnitude
  private var maxExecutionTime: Double = 0.0

  var hasEnoughData: Bool {
    return performanceSamplesQueue.sync {
      return performanceSamples.count >= warmupSamples
    }
  }

  var averageExecutionTimeSeconds: Double {
    return performanceSamplesQueue.sync {
      guard !performanceSamples.isEmpty else { return 0.0 }

      // Skip warmup samples for more accurate measurement
      let validSamples = performanceSamples.count > warmupSamples
        ? Array(performanceSamples.dropFirst(warmupSamples))
        : performanceSamples

      guard !validSamples.isEmpty else { return 0.0 }

      // Use trimmed mean to reduce impact of outliers
      let sorted = validSamples.sorted()
      let trimPercent = 0.1 // Remove top and bottom 10%
      let trimCount = Int(Double(sorted.count) * trimPercent)

      let trimmed = (trimCount > 0 && sorted.count > trimCount * 2)
        ? Array(sorted.dropFirst(trimCount).dropLast(trimCount))
        : sorted

      let sum = trimmed.reduce(0, +)
      let average = sum / Double(trimmed.count)

      lastEvaluation = counter
      return average
    }
  }

  var processingStats: [String: Any] {
    return performanceSamplesQueue.sync {
      return [
        "totalFrames": totalFramesProcessed,
        "totalTime": Double(totalProcessingTime) / 1_000_000_000.0, // Convert to seconds
        "averageTime": averageExecutionTimeSeconds,
        "minTime": minExecutionTime,
        "maxTime": maxExecutionTime,
        "sampleCount": performanceSamples.count
      ]
    }
  }

  func beginPerformanceSampleCollection() -> PerformanceSampleCollection {
    let begin = DispatchTime.now()

    return PerformanceSampleCollection {
      let end = DispatchTime.now()
      let nanoseconds = end.uptimeNanoseconds - begin.uptimeNanoseconds
      let seconds = Double(nanoseconds) / 1_000_000_000.0

      self.performanceSamplesQueue.async {
        let index = self.counter % maxSampleSize

        if self.performanceSamples.count > index {
          self.performanceSamples[index] = seconds
        } else {
          self.performanceSamples.append(seconds)
        }

        // Update statistics
        self.totalFramesProcessed += 1
        self.totalProcessingTime += nanoseconds
        self.minExecutionTime = min(self.minExecutionTime, seconds)
        self.maxExecutionTime = max(self.maxExecutionTime, seconds)

        self.counter += 1
      }
    }
  }

  func clear() {
    performanceSamplesQueue.sync {
      counter = 0
      performanceSamples.removeAll()
      totalFramesProcessed = 0
      totalProcessingTime = 0
      minExecutionTime = Double.greatestFiniteMagnitude
      maxExecutionTime = 0.0
    }
  }

  func getSuggestedFrameRate(maxFrameRate: Double = 30.0) -> Double {
    guard hasEnoughData else { return maxFrameRate }

    let avgTime = averageExecutionTimeSeconds
    guard avgTime > 0 else { return maxFrameRate }

    // Calculate theoretical max FPS based on processing time
    let theoreticalMaxFps = 1.0 / avgTime

    // Add safety margin (20%) to prevent frame drops
    let safeMaxFps = theoreticalMaxFps * 0.8

    // Clamp to reasonable bounds
    return min(safeMaxFps, maxFrameRate).clamped(to: 1.0...maxFrameRate)
  }
}

// MARK: - Double Extension

private extension Double {
  func clamped(to range: ClosedRange<Double>) -> Double {
    return Swift.min(Swift.max(self, range.lowerBound), range.upperBound)
  }
}
